## Assignment 3 (One) 
Polli Fakhretdinova

=======================================================================================================================

In [2]:
import pandas as pd
import numpy as np
import sys as s

In [3]:
df = pd.read_csv('olympics.csv', index_col=0, skiprows=1)

for col in df.columns:
    if col[:2]=='01':
        df.rename(columns={col:'Gold'+col[4:]}, inplace=True)
    if col[:2]=='02':
        df.rename(columns={col:'Silver'+col[4:]}, inplace=True)
    if col[:2]=='03':
        df.rename(columns={col:'Bronze'+col[4:]}, inplace=True)
    if col[:1]=='№':
        df.rename(columns={col:'#'+col[1:]}, inplace=True)

names_ids = df.index.str.split('\s\(')

df.index = names_ids.str[0] 

In [4]:
df['ID'] = names_ids.str[1]
df = df.drop('Totals')

======================================================================================================================

## Part I: Use Olympics Data to Answer the Next 4 Questions

What does the content of the olympics data look like?

In [5]:
def content_olympics ():
    
    return df.head(4)

content_olympics () 

№ Summer  Gold  Silver  Bronze  Total  № Winter  Gold.1  \
Afghanistan (AFG)        13     0       0       2      2         0       0   
Algeria (ALG)            12     5       2       8     15         3       0   
Argentina (ARG)          23    18      24      28     70        18       0   
Armenia (ARM)             5     1       2       9     12         6       0   

                   Silver.1  Bronze.1  Total.1  № Games  Gold.2  Silver.2  \
Afghanistan (AFG)         0         0        0       13       0         0   
Algeria (ALG)             0         0        0       15       5         2   
Argentina (ARG)           0         0        0       41      18        24   
Armenia (ARM)             0         0        0       11       1         2   

                   Bronze.2  Combined total  ID  
Afghanistan (AFG)         2               2 NaN  
Algeria (ALG)             8              15 NaN  
Argentina (ARG)          28              70 NaN  
Armenia (ARM)             9              12 NaN

=======================================================================================================================

### Question 1

Which country has won the most gold medals in summer games?
    
This function should return a single string value.

In [6]:
def answer_one():
    
    df_a = summer = df['Gold']
    
    return summer[summer == max(summer)]

answer_one()

United States (USA) [P] [Q] [R] [Z]    976
Name: Gold, dtype: int64

Answer 1: 

United States has won the most gold medals in summer games. In total, USA has won 976 gold medals in summer games.

=======================================================================================================================

### Question 2

Which country had the biggest difference between their summer and winter gold medal counts?
    
This function should return a single string value.

In [7]:
def answer_two():
    
    df_x = summer = df['Gold']
    df_x = winter = df['Gold.1']
    
    df_x = difference = summer - winter
    
    return difference[difference == max(difference)]
answer_two()

United States (USA) [P] [Q] [R] [Z]    880
dtype: int64

Answer 2: 

United States had the biggest difference between its summer and winter gold medal counts. USA won 880 more gold medals in summer games than in winter games.

=======================================================================================================================

### Question 3

Which country has the biggest difference between their summer gold medal counts and winter gold medal counts relative to their total gold medal count? --> Summer Gold − Winter Gold / Total Gold
 
Only include countries that have won at least 1 gold in both summer and winter.
This function should return a single string value.

In [9]:
def answer_three():
    
    df_p = df[df['Gold'] > 0]
    df_pp = df_p[df_p['Gold.1'] > 0]
    
    df_f = summer = df_pp['Gold']
    df_f = winter = df_pp['Gold.1']
    
    df_f = total = summer + winter
    df_f = difference = summer - winter
    df_f = percent = (difference / total) * 100
    
    return percent[percent == max(percent)]

answer_three()

Bulgaria (BUL) [H]    96.153846
dtype: float64

Answer 3: 

Bulgaria has the biggest difference between its summer gold medal counts and winter gold medal counts relative to their total gold medal count, with the 96.15%. 

=======================================================================================================================

### Question 4

Write a function that creates a Series called "Points" which is a weighted value where each gold medal (Gold.2) counts for 3 points, silver medals (Silver.2) for 2 points, and bronze medals (Bronze.2) for 1 point. 

The function should return only the column (a Series object) which you created.

In [9]:
def answer_four():
    
    Gold = df['Gold.2'] * 3
    Silver = df['Silver.2'] * 2
    Bronze = df['Bronze.2'] * 1
    
    Points = Gold + Silver + Bronze
    
    return Points

answer_four()

Afghanistan (AFG)                                  2
Algeria (ALG)                                     27
Argentina (ARG)                                  130
Armenia (ARM)                                     16
Australasia (ANZ) [ANZ]                           22
Australia (AUS) [AUS] [Z]                        923
Austria (AUT)                                    569
Azerbaijan (AZE)                                  43
Bahamas (BAH)                                     24
Bahrain (BRN)                                      1
Barbados (BAR) [BAR]                               1
Belarus (BLR)                                    154
Belgium (BEL)                                    276
Bermuda (BER)                                      1
Bohemia (BOH) [BOH] [Z]                            5
Botswana (BOT)                                     2
Brazil (BRA)                                     184
British West Indies (BWI) [BWI]                    2
Bulgaria (BUL) [H]                            

======================================================================================================================

## Part II: Use the Census Data Frame to Answer the Next 4 Questions

In [10]:
c_df = pd.read_csv('census.csv')

What does the content of the census data look like?

In [11]:
def content_census ():
    
    return c_df.head(4)

content_census () 

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010     ...       \
0        4779736            4780127          4785161     ...        
1          54571              54571            54660     ...        
2         182265             182265           183193     ...        
3          27457              27457            27341     ...        

   RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0          0.002295         -0.193196          0.381066          0.582002   
1          7.242091         -2.915927         -3.012349          2.265971   
2         14.832960         17.647293         21.845705         19.243287   
3         -4.728132         -2.500690         -7.056824         -3.904217   

   RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0         -0.467369     1.030015     0.826644     1.383282     1.724718   
1         -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2         17.197872    15.844176    18.559627    22.727626    20.317142   
3        -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   

   RNETMIG2015  
0     0.712594  
1    -2.187333  
2    18.293499  
3   -10.543299  

[4 rows x 100 columns]

======================================================================================================================

### Question 5

Which state has the most counties in it? 

(hint: consider the sumlevel key carefully! You'll need this for future questions too...)

In [13]:
def answer_five():
    
    grouped = c_df.groupby(["STNAME"],sort=False).sum()["COUNTY"].idxmax()
    

    return grouped

answer_five()

'Texas'

Answer 5: State of Texas has the most counties in it.

======================================================================================================================

### Question 6

Only looking at the three most populous counties for each state, what are the three most populous states (in order of highest population to lowest population)?  

Use CENSUS2010POP.

In [14]:
def answer_six():
    
    p0 = c_df[c_df['SUMLEV'] == 50].groupby('STNAME')['CENSUS2010POP'].nlargest(2)
    
    p1 = p0.nlargest(3)
    
    return p1

answer_six()

STNAME          
California  209     9818605
Illinois    624     5194675
Texas       2667    4092459
Name: CENSUS2010POP, dtype: int64

Answer 6: 

State of California, Illinois, and Texas are the three most populous states.

======================================================================================================================

### Question 7

Which county has had the largest absolute change in population within the period 2010-2015? 

(Hint: population values are stored in columns POPESTIMATE2010 through POPESTIMATE2015, you need to consider all six columns.)

e.g. If County Population in the 5 year period is 100, 120, 80, 105, 100, 130, then its largest change in the period would be |130-80| = 50.

In [15]:
def answer_seven():
    
    k = c_df[c_df['SUMLEV'] == 50]
    
    k0 = k.set_index(['STNAME','CTYNAME'])
    
    k00 = k0.loc[:,['POPESTIMATE2010',
                    'POPESTIMATE2011',
                    'POPESTIMATE2012',
                    'POPESTIMATE2013',
                    'POPESTIMATE2014',
                    'POPESTIMATE2015']].stack()
    
    k1 = k00.max(level=['STNAME','CTYNAME']) 
    k2 = k00.min(level=['STNAME','CTYNAME'])
    k3 = k1 - k2
    
    return k3[k3 == max(k3)]   

answer_seven()

STNAME  CTYNAME      
Texas   Harris County    429841
dtype: int64

Answer 7: 

The Harris County of Texas has had the largest absolute change in population within the period 2010-2015, which equals to 429, 841.

======================================================================================================================

### Question 8 

In this datafile, the United States is broken up into four regions using the "REGION" column.

Create a query that finds the counties that belong to regions 1 or 2, whose name starts with 'Washington', and whose POPESTIMATE2015 was greater than their POPESTIMATE 2014.

This function should return a 5x2 DataFrame with the columns = ['STNAME', 'CTYNAME'] and the same index ID as the census_df (sorted ascending by index).

In [16]:
def answer_eight():
    
    f1 = c_df[(c_df['POPESTIMATE2015'] > c_df['POPESTIMATE2014']) & 
              
              ((c_df['REGION'] == 1) | (c_df['REGION'] == 2))]
    
    f1 = f1[f1['CTYNAME'].str.startswith('Washington')]
    
    f5 = f1.loc[:,['STNAME', 'CTYNAME']]
    
    f2 = f5.sort_values('STNAME',ascending = True)
    
    return f2

answer_eight()

STNAME            CTYNAME
896           Iowa  Washington County
1419     Minnesota  Washington County
2345  Pennsylvania  Washington County
2355  Rhode Island  Washington County
3163     Wisconsin  Washington County

Answer 8: 

The counties that belong to regions 1 or 2, whose name starts with 'Washington', and whose POPESTIMATE2015 was greater than their POPESTIMATE 2014 are located in states: Iowa, Minnesota, Pennsylvania, Rhode Island, and Wisconsin.

======================================================================================================================